### PR TITLE
[CALCITE-5127] Error when executing query with subquery in select lis…

### PR DESCRIPTION
**Pull request by libenchao**
…t that uses outer column of array type

Root cause of this issue is that currently Projection does not support correlation, see https://issues.apache.org/jira/browse/CALCITE-1045
In this pr, I added support for correlation in Projection.

### CODE REVIEW

**Review by dssysolyatin — APPROVED** [review]
Looks good for me

**Comment by ian-bertolacci** [comment]
Hello, what is the status of this PR?

**Comment by libenchao** [comment]
@ian-bertolacci This pr was in discussion in JIRA, and I think we have reached an agreement.
Since we are about releasing 1.31.0, I didn't push this too much because we decided to move it to 1.32.0 (I'll try to push this forward once 1.31.0 was released)

**Comment by zabetak** [comment]
The PR is in very good shape, I promise to make a full pass during next week and hopefully merge it as well. Apologies for the delay, most of it was due to me.

**Comment by ian-bertolacci** [comment]
@libenchao @zabetak thanks for the update! Good to hear this is in a good and ready state. My understanding is that 1.31 has been delayed by other issues; is there a possiblity that this can sneak into the 1.31 while those other issues are resolved?

**Review by libenchao — COMMENTED** [review]
@zabetak Thanks for the review, and sorry for the late response. 
I've rebased to resolve conflicts, and address most of the comments.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 1]
```diff
@@ -1908,9 +1927,11 @@ private RelBuilder project_(
       fieldNameList.add(null);
     }
 
+    // Do not merge projection when top projection has correlation variables
     bloat:
     if (frame.rel instanceof Project
-        && config.bloat() >= 0) {
+        && config.bloat() >= 0
+        && variables.isEmpty()) {
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/tools/RelBuilder.java:1934]: Based on my latest comments under the JIRA about not adding variablesSet in `Project`, I was thinking that something like the snippet below could be used here and similarly in other places.

```java
if (StreamSupport.stream(nodes.spliterator(), false).anyMatch(RexUtil::containsCorrelation)) {
  break bloat;
}
```

**Comment by libenchao** [comment]
https://github.com/apache/calcite/pull/2911 has been opened, so I'm closing this now.

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProjectRule.java [hunk 2]
```diff
@@ -42,6 +43,14 @@ protected EnumerableProjectRule(Config config) {
     super(config);
   }
 
+  @Override public boolean matches(RelOptRuleCall call) {
+    LogicalProject project = call.rel(0);
+    if (!project.getVariablesSet().isEmpty()) {
+      return false;
+    }
+    return true;
```
Comment (zabetak) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumerableProjectRule.java:51]: Replace with `return project.getVariablesSet().isEmpty();` Idem to the other rules.

File: core/src/main/java/org/apache/calcite/tools/RelBuilder.java [hunk 3]
```diff
@@ -1817,7 +1817,24 @@ public RelBuilder project(Iterable<? extends RexNode> nodes,
...
+   * @param fieldNames Suggested field names
+   * @param force create project even if it is identity
+   * @param variablesSet Correlating variables that are set when reading a row
+   *                     from the input, and which may be referenced from the
+   *                     projection expressions
+   */
+  public RelBuilder project(Iterable<? extends RexNode> nodes,
+      Iterable<? extends @Nullable String> fieldNames, boolean force,
+      Iterable<CorrelationId> variablesSet) {
+    return project_(nodes, fieldNames, ImmutableList.of(), force, variablesSet);
```
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/tools/RelBuilder.java:1837]: There are certain disadvantages with the approach I suggested. We can take advantage of `variablesSet` in various places such as `RelFieldTrimmer` so overall it makes sense to move forward with this change. Marking this as resovled.

File: core/src/test/java/org/apache/calcite/plan/RelWriterTest.java [hunk 4]
```diff
@@ -998,6 +999,21 @@ void testCorrelateQuery(SqlExplainFormat format) {
...
+    final Function<RelBuilder, RelNode> relFn = b -> b.scan("EMP")
+        .project(
+            ImmutableList.of(b.field("ENAME")),
+            ImmutableList.of("ename"),
+            true,
+            ImmutableSet.of(b.getCluster().createCorrel()))
+        .build();
+
+    final String expected = "LogicalProject(variablesSet=[[$cor0]], ename=[$1])\n"
+        + "  LogicalTableScan(table=[[scott, EMP]])\n";
```
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/test/java/org/apache/calcite/plan/RelWriterTest.java:1012]: Agreed!

File: core/src/main/java/org/apache/calcite/rel/core/Project.java [hunk 5]
```diff
@@ -137,20 +163,27 @@ protected Project(RelInput input) {
...
    */
   public abstract Project copy(RelTraitSet traitSet, RelNode input,
-      List<RexNode> projects, RelDataType rowType);
+      List<RexNode> projects, RelDataType rowType, Set<CorrelationId> variableSet);
+
+  @Deprecated // to be removed before 2.0
+  public Project copy(RelTraitSet traitSet, RelNode input,
+      List<RexNode> projects, RelDataType rowType) {
+    return copy(traitSet, input, projects, rowType, ImmutableSet.of());
+  }
```
[the remark this answers is not in the document]
Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/core/Project.java:180]: The new copy method is used, such as following places: `FilterProjectTransposeRule`, `ProjectJoinRemoveRule`, `ProjectWindowTransposeRule`, `RelBuilder`. 
And I still keep the old copy method to make it a compatible change. (We already changed `ProjectFactory`, hence I think it's ok to also add a new version copy method to allow downstream projects to adapt to the new `Project` with variables)

**Review by libenchao — COMMENTED** [review]
@zabetak I addressed and replied to your last comments (if I missed any, please ping me). 
And I also rebased to the latest main branch to make sure the PR is not broken by any new added codes.

File: core/src/main/java/org/apache/calcite/rel/core/Project.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zabetak) [reply core/src/main/java/org/apache/calcite/rel/core/Project.java:180]: The question is if it is used in a meaningful way? As far as I can see we are always passing the same `variableSet`. If we don't need to change the `variableSet` as part of the copy then we don't need the new method.

[the remark this answers is not in the document]
Comment (libenchao) [reply core/src/main/java/org/apache/calcite/rel/core/Project.java:180]: Sounds reasonable, I'll remove the new copy method.

File: mongodb/src/main/java/org/apache/calcite/adapter/mongodb/MongoProject.java [hunk 6]
```diff
@@ -57,8 +58,7 @@ public MongoProject(RelOptCluster cluster, RelTraitSet traitSet,
 
   @Override public Project copy(RelTraitSet traitSet, RelNode input,
       List<RexNode> projects, RelDataType rowType) {
-    return new MongoProject(getCluster(), traitSet, input, projects,
-        rowType);
+    return new MongoProject(getCluster(), traitSet, input, projects, rowType);
```
Comment (divyanshuS27) [inline mongodb/src/main/java/org/apache/calcite/adapter/mongodb/MongoProject.java:61]: don't we need any specific test cases for modules other than core module that are impacted?

**Pull request closed without merging**